rust/template/d3log/src/value_set.rs

```diff
@@ -0,0 +1,117 @@
+// module to wrap the general encoding type DDValue for rust compiled ddlog
+
+use crate::{Batch, BatchBody, Error, Evaluator};
+use differential_datalog::{ddval::DDValue, program::RelId, DeltaMap};
+use std::collections::{BTreeMap, HashMap};
+use std::fmt;
+use std::fmt::Display;
+use std::sync::{Arc, Mutex};
+
+#[derive(Clone)]
+pub struct ValueSet {
+    pub eval: Evaluator, // used to translate to record
+    pub deltas: Arc<Mutex<DeltaMap<differential_datalog::ddval::DDValue>>>,
+}
+
+impl Display for ValueSet {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&"<")?;
+        let iter = self.clone();
+        let mut m: usize = 0;
+        for (relid, _v, _w) in iter.into_iter() {
+            // I would really prefer a readable name..but we need an Evaluator and it doesn't seem right
+            // to add it to all the batches or globalize it
+            f.write_str(&format!("({}", relid))?;
+            m += 1;
+            f.write_str(&format!(" {})", m))?;
+        }
+        f.write_str(&">")?;
+        Ok(())
+    }
+}
+
+pub struct BatchIterator<'a> {
+    relid: RelId,
+    relations: Box<dyn Iterator<Item = (RelId, BTreeMap<DDValue, isize>)> + Send + 'a>,
+    items: Option<Box<dyn Iterator<Item = (DDValue, isize)> + Send>>,
+}
+
+impl<'a> Iterator for BatchIterator<'a> {
+    type Item = (RelId, DDValue, isize);
+
+    fn next(&mut self) -> Option<(RelId, DDValue, isize)> {
+        match &mut self.items {
+            Some(x) => match x.next() {
+                Some((v, w)) => Some((self.relid, v, w)),
+                None => {
+                    self.items = None;
+                    self.next()
+                }
+            },
+            None => {
+                // what about the empty batch?
+                let (relid, items) = self.relations.next()?;
+                self.relid = relid;
+                self.items = Some(Box::new(items.into_iter()));
+                self.next()
+            }
+        }
+    }
```

Comment:
This seems really fragile and potentially infinitely recursive if the `self.next()` invocations keep cycling